x/subkeys/internal/keeper/keeper.go

```diff
@@ -0,0 +1,139 @@
+package keeper
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/subkeys/exported"
+	"github.com/cosmos/cosmos-sdk/x/subkeys/internal/types"
+)
+
+type Keeper struct {
+	cdc      *codec.Codec
+	storeKey sdk.StoreKey
+}
+
+// NewKeeper creates a DelegationKeeper
+func NewKeeper(cdc *codec.Codec, storeKey sdk.StoreKey) Keeper {
+	return Keeper{cdc: cdc, storeKey: storeKey}
+}
+
+// DelegateFeeAllowance creates a new grant
+func (k Keeper) DelegateFeeAllowance(ctx sdk.Context, grant types.FeeAllowanceGrant) error {
+	store := ctx.KVStore(k.storeKey)
+	key := types.FeeAllowanceKey(grant.Granter, grant.Grantee)
+
+	bz, err := k.cdc.MarshalBinaryBare(grant)
+	if err != nil {
+		return err
+	}
+	store.Set(key, bz)
+	return nil
+}
+
+// RevokeFeeAllowance removes an existing grant
+func (k Keeper) RevokeFeeAllowance(ctx sdk.Context, granter sdk.AccAddress, grantee sdk.AccAddress) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.FeeAllowanceKey(granter, grantee)
+	store.Delete(key)
+}
+
+// GetFeeAllowance returns the allowance between the granter and grantee.
+// If there is none, it returns nil, nil.
+// Returns an error on parsing issues
+func (k Keeper) GetFeeAllowance(ctx sdk.Context, granter sdk.AccAddress, grantee sdk.AccAddress) (exported.FeeAllowance, error) {
+	grant, err := k.GetFeeGrant(ctx, granter, grantee)
+	if grant == nil {
+		return nil, err
+	}
+	return grant.Allowance, err
+}
+
+// GetFeeGrant returns entire grant between both accounts
+func (k Keeper) GetFeeGrant(ctx sdk.Context, granter sdk.AccAddress, grantee sdk.AccAddress) (*types.FeeAllowanceGrant, error) {
```

Comment:
In my experience with standard go libraries, there is an idiom when returning an error result, that is should be `return 0, err` or `return nil, err`. I could return the zero-value of grant, but that seems ugly and non-standard to me. I have seen `return SomeStruct{}, err` mainly in cosmos-sdk code, and then most of that starts using named return values to avoid declaring this object, and adds an `IsZero` method. While this works, I prefer to stick with what I consider more idiomatic go (in the larger language community).

The only downside I see is a bit of work for the garbage collector, but this isn't rust. (Where I would pass `Result<Grant, Error>` and do it by value anyway.

Reply:
I see. Yes, we typically return the zero value struct and solely rely on if an error is returned or not. I don't see this as a problem here:

```go
grant, err := k.GetFeeGrant(ctx, granter, grantee)
if err != nil {
  // handle error
}

// at this point, you can assume grant is valid
```

`GetFeeGrant` can handle the case if the grant is actually `nil` and return an error. If you feel strongly about this, we can leave it.